controllers/controllers/workloads/integration/cfspace_controller_integration_test.go

```diff
@@ -295,25 +296,127 @@ var _ = Describe("CFSpaceReconciler Integration Tests", func() {
+			rootServiceAccount.Labels = map[string]string{"new-label": "dummy-value"}
 			Expect(k8sClient.Patch(ctx, rootServiceAccount, client.MergeFrom(origRootServiceAccount))).To(Succeed())
 		})
 
 		It("doesn't modify the secrets on the propagated service account", func() {
-			Consistently(func(g Gomega) {
+			Eventually(func(g Gomega) {
+				var updatedPropagatedServiceAccount corev1.ServiceAccount
+				g.Expect(
+					k8sClient.Get(ctx, client.ObjectKeyFromObject(&propagatedServiceAccount), &updatedPropagatedServiceAccount),
+				).To(Succeed())
+				g.Expect(updatedPropagatedServiceAccount.Secrets).To(ConsistOf(
+					corev1.ObjectReference{Name: tokenSecretName},
+					corev1.ObjectReference{Name: packageRegistrySecretName},
+				))
+				g.Expect(updatedPropagatedServiceAccount.ImagePullSecrets).To(Equal(rootServiceAccount.ImagePullSecrets))
+			}).Should(Succeed())
+		})
+	})
+
+	When("the package registry secret is added to the root service account", func() {
+		var (
+			rootServiceAccount       *corev1.ServiceAccount
+			propagatedServiceAccount corev1.ServiceAccount
+			tokenSecretName          string
+		)
+		BeforeEach(func() {
+			Expect(k8sClient.Create(ctx, cfSpace)).To(Succeed())
+			Eventually(func(g Gomega) {
+				var createdSpace korifiv1alpha1.CFSpace
+				g.Expect(k8sClient.Get(ctx, types.NamespacedName{Namespace: orgNamespace.Name, Name: spaceGUID}, &createdSpace)).To(Succeed())
+				g.Expect(meta.IsStatusConditionTrue(createdSpace.Status.Conditions, "Ready")).To(BeTrue())
+			}, 20*time.Second).Should(Succeed())
+
+			rootServiceAccount = &corev1.ServiceAccount{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:        PrefixedGUID("existing-service-account"),
+					Namespace:   cfRootNamespace,
+					Annotations: map[string]string{"cloudfoundry.org/propagate-service-account": "true"},
+				},
+			}
+			Expect(k8sClient.Create(ctx, rootServiceAccount)).To(Succeed())
+
+			// Ensure that the service account is propagated into the CFSpace namespace
+			Eventually(func() error {
+				return k8sClient.Get(ctx, types.NamespacedName{Name: rootServiceAccount.Name, Namespace: cfSpace.Name}, &propagatedServiceAccount)
+			}).Should(Succeed())
+
+			// Simulate k8s adding a token secret to the propagated service account
+			origPropagatedServiceAccount := propagatedServiceAccount.DeepCopy()
+			tokenSecretName = rootServiceAccount.Name + "-token-XYZABC"
+			propagatedServiceAccount.Secrets = []corev1.ObjectReference{{Name: tokenSecretName}}
+			Expect(k8sClient.Patch(ctx, &propagatedServiceAccount, client.MergeFrom(origPropagatedServiceAccount))).To(Succeed())
+
+			// Add the package registry secret to the root service account
+			origRootServiceAccount := rootServiceAccount.DeepCopy()
+			rootServiceAccount.Secrets = []corev1.ObjectReference{{Name: packageRegistrySecretName}}
+			Expect(k8sClient.Patch(ctx, rootServiceAccount, client.MergeFrom(origRootServiceAccount))).To(Succeed())
```

Comment:
We have recently introduced [`k8s.PatchResource`](https://github.com/cloudfoundry/korifi/blob/72cc20246f06fe67741da88d24364ac6f5fb6cec/tools/k8s/patch.go#L80) (or `Patch` if you want to also patch status) to patch k8s objects so that you do not have to do the `deep-copy` - `client.MergeFrom` dance yourself. We spent quite some time on using these utilities all across the codebase (we might have missed some places though) to simplify the code a bit. I would suggest using them here as well.